src/nitro/src/enclaves.rs

```diff
@@ -1,97 +1,279 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use super::error::NitroError;
+use flate2::read::GzDecoder;
+use libc::c_int;
 use nitro_enclaves::{
     launch::{ImageType, Launcher, MemoryInfo, PollTimeout, StartFlags},
     Device,
 };
 use nix::poll::{poll, PollFd, PollFlags, PollTimeout as NixPollTimeout};
 use std::{
-    fs::File,
-    io::{Read, Write},
+    ffi::CString,
+    fs,
+    io::{self, ErrorKind, Read, Write},
     os::fd::AsFd,
+    path::PathBuf,
+    str::FromStr,
 };
-use vsock::{VsockAddr, VsockListener};
+use tar::{Archive, HeaderMode};
+use vsock::{VsockAddr, VsockListener, VsockStream};
 
 type Result<T> = std::result::Result<T, NitroError>;
 
+const KRUN_NITRO_EIF_TAR: &[u8] = include_bytes!("runtime-data/eif.tar.gz");
+const KRUN_NITRO_EIF_FILE_NAME: &str = "krun-nitro.eif";
+
 const ENCLAVE_READY_VSOCK_PORT: u32 = 9000;
 
 const VMADDR_CID_PARENT: u32 = 3;
 
 const HEART_BEAT: u8 = 0xb7;
 
+const ROOTFS_DIR_DENYLIST: [&str; 5] = [
+    "proc", // /proc.
+    "run",  // /run.
+    "tmp",  // /tmp.
+    "dev",  // /dev.
+    "sys",  // /sys.
+];
+
 /// Nitro Enclave data.
 pub struct NitroEnclave {
-    /// Enclave image.
-    pub image: File,
+    /// Path to configurable enclave image. Will default to KRUN_NITRO_ENCLAVE_EIF if one external
+    /// enclave provided.
+    pub _image_path: Option<PathBuf>,
     /// Amount of RAM (in MiB).
     pub mem_size_mib: usize,
     /// Number of vCPUs.
     pub vcpus: u8,
+    /// Enclave rootfs.
+    pub rootfs: String,
     /// Enclave start flags.
     pub start_flags: StartFlags,
+    /// Execution path.
+    pub exec_path: String,
+    /// Execution args.
+    pub exec_args: String,
+    /// Execution environment.
+    pub exec_env: String,
 }
 
 impl NitroEnclave {
     /// Run the enclave.
     pub fn run(&mut self) -> Result<u32> {
+        let rootfs_archive = self.rootfs_archive()?;
+
+        let (cid, mut stream) = self.start()?;
+
+        vsock_write_bytes(&rootfs_archive, &mut stream)?;
+
+        self.write_exec(&mut stream)?;
+
+        Ok(cid)
+    }
+
+    fn rootfs_archive(&self) -> Result<Vec<u8>> {
+        let mut builder = tar::Builder::new(Vec::new());
+
+        builder.mode(HeaderMode::Deterministic);
+        builder.follow_symlinks(false);
+
+        let rootfs_dirname: String = {
+            let split: Vec<&str> = self.rootfs.split("/").collect();
+
+            split.last().unwrap_or(&"/").to_string()
+        };
+
+        for entry in
+            fs::read_dir(PathBuf::from(self.rootfs.clone())).map_err(NitroError::RootFsArchive)?
+        {
```

Comment:
I think you can do something like
```suggestion
        let pathbuf = PathBuf::from(self.rootfs.clone());
        let rootfs_dirname = pathbuf.file_name().unwrap_or(&OsStr::new("/")).to_str().unwrap();
        
        for entry in
            fs::read_dir(pathbuf).map_err(NitroError::RootFsArchive)?
        {
```